src/spark_rapids_ml/knn.py

```diff
@@ -235,54 +210,60 @@ def setK(self, value: int) -> "NearestNeighbors":
-        if not self.isDefined("id_col"):
-            query_df = self._df_zip_with_index(query_df)
-        processed_query_df = query_df.withColumn(alias.label, lit(self.label_isquery))
 
-        union_df = self.processed_item_df.union(processed_query_df)
-
-        pipelinedrdd = self._fit(union_df)
-        pipelinedrdd = pipelinedrdd.repartition(query_default_num_partitions)  # type: ignore
-        knn_rdd = pipelinedrdd.flatMap(
-            lambda row: list(zip(row["query_id"], row["indices"], row["distances"]))
+        model = self._create_pyspark_model(
+            Row(
+                item_df=self.item_df,
+                processed_item_df=self.processed_item_df,
+                label_isdata=self.label_isdata,
+                label_isquery=self.label_isquery,
+            )
         )
-        knn_df = knn_rdd.toDF(
-            schema=f"query_{self.getIdCol()} int, indices array<int>, distances array<float>"
-        ).sort(f"query_{self.getIdCol()}")
+        model._num_workers = self._num_workers
+        self._copyValues(model)
+        self._copy_cuml_params(model)  # type: ignore
+        return model
 
-        return (query_df, self.item_df, knn_df)
+    def _out_schema(self) -> Union[StructType, str]:  # type: ignore
+        """
+        This class overrides _fit and will not call _out_schema.
+        """
+        pass
 
-    def _df_zip_with_index(self, df: DataFrame) -> DataFrame:
+    def _get_cuml_fit_func(  # type: ignore
+        self, dataset: DataFrame
+    ) -> Callable[[CumlInputType, Dict[str, Any]], Dict[str, Any],]:
         """
-        Add an row number column (or equivalently id column) to df using zipWithIndex. Used when id_col is not set.
-        TODO: May replace zipWithIndex with monotonically_increasing_id if row number does not have to consecutive.
+        This class overrides _fit and will not call _get_cuml_fit_func.
         """
-        out_schema = StructType(
-            [StructField(self.row_number_col, IntegerType(), False)] + df.schema.fields
-        )
-        zipped_rdd = df.rdd.zipWithIndex()
-        new_rdd = zipped_rdd.map(lambda row: [row[1]] + list(row[0]))
-        return new_rdd.toDF(schema=out_schema)
+        pass
 
-    def _require_ucx(self) -> bool:
-        return True
 
-    def _return_model(self) -> bool:
-        return False
+class NearestNeighborsModel(
+    _CumlCaller, _CumlModel, NearestNeighborsClass, _NearestNeighborsCumlParams
+):
+    def __init__(
+        self,
+        item_df: DataFrame,
+        processed_item_df: DataFrame,
+        label_isdata: int,
+        label_isquery: int,
+    ):
+        super().__init__()
+        self.item_df = item_df
+        self.processed_item_df = processed_item_df
+        self.label_isdata = label_isdata
+        self.label_isquery = label_isquery
 
-    def _out_schema(self) -> Union[StructType, str]:
+    def _out_schema(self) -> Union[StructType, str]:  # type: ignore
```

Comment:
do we still need `# type: ignore` here?

Reply:
Yeah. This is because _CumlCaller has an abstractmethod _out_schema(self), and _CumlModel also has an abstractmethod _out_schema(self, input_schema: StructType).  

Was not able to implement both methods in NearestNeighborsModel, yet uncomment the # type: ignore will trigger a mypy error.